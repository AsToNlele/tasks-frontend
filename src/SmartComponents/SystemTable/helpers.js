const buildSortString = (orderBy, orderDirection) => {
  let sortString = orderBy ? '&sort=' : '';
  let direction = '';
  if (orderDirection === 'DESC') {
    direction = '-';
  }
  let order = orderBy === 'updated' ? 'last_seen' : orderBy;

  return `${sortString}${direction}${order}`;
};

const buildFilterString = (filters) => {
  let displayNameFilter = filters.hostnameOrId
    ? `&display_name=${filters.hostnameOrId}`
    : '';

  let osFilter = '';
  for (const majorVer in filters.osFilter) {
    for (const majMinVer in filters.osFilter[majorVer]) {
      if (filters.osFilter[majorVer][majMinVer]) {
        osFilter += `&os_version=${majMinVer}`;
      }
    }
  }

  return `${displayNameFilter}${osFilter}`;
};

export const buildFilterSortString = (
  limit,
  offset,
  orderBy,
  orderDirection,
  filters
) => {
  let limitOffsetString = `limit=${limit}&offset=${offset}`;
  let sortString = buildSortString(orderBy, orderDirection);
  let filterString = buildFilterString(filters);
  return `?${limitOffsetString}${sortString}${filterString}`;
};

export const findCheckedValue = (total, selected) => {
  if (selected === total && total > 0) {
    return true;
  } else if (selected > 0 && selected < total) {
    return null;
  } else {
    return false;
  }
};
