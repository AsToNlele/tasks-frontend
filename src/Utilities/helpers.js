import React from 'react';
import moment from 'moment';
import {
  Skeleton,
  SkeletonSize,
} from '@redhat-cloud-services/frontend-components/Skeleton';

const renderRunning = (time) => {
  return time === 'null' || !time ? 'Running' : false;
};

export const renderRunDateTime = (time) => {
  if (time === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return (
      renderRunning(time) || moment.utc(time).format('DD MMM YYYY, HH:mm UTC')
    );
  }
};

export const getTimeDiff = ([start, end]) => {
  if (start === 'loading') {
    return <Skeleton size={SkeletonSize.md} />;
  } else {
    return (
      renderRunning(end) ||
      `${renderRunDateTime(end)} (${moment
        .duration(
          moment(renderRunDateTime(end), 'DD MMM YYYY, HH:mm').diff(
            moment(renderRunDateTime(start), 'DD MMM YYYY HH:mm')
          )
        )
        .asMinutes()} min)`
    );
  }
};

export const uniq = (collection) => [...new Set(collection)];

/*eslint-disable react/display-name*/
export const renderColumnComponent =
  (Component, props) => (_data, _id, entity) =>
    <Component {...entity} {...props} />;
/*eslint-enable react/display-name*/

export const orderArrayByProp = (property, objects, direction) =>
  objects.sort((a, b) => {
    if (direction === 'asc') {
      if (typeof a[property] === 'number') {
        return a[property] - b[property];
      } else {
        return String(a[property]).localeCompare(String(b[property]));
      }
    } else {
      if (typeof a[property] === 'number') {
        return -a[property] + b[property];
      } else {
        return -String(a[property]).localeCompare(String(b[property]));
      }
    }
  });

export const getProperty = (obj, path, fallback) => {
  const parts = path.split('.');
  const key = parts.shift();
  if (typeof obj[key] !== 'undefined') {
    return parts.length > 0
      ? getProperty(obj[key], parts.join('.'), fallback)
      : obj[key];
  }

  return fallback;
};

export const camelCase = (string) => {
  return string
    .split(/[-_\W]+/g)
    .map((string) => string.trim())
    .map((string) => string[0].toUpperCase() + string.substring(1))
    .join('');
};
