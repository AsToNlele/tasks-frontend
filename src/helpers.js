import React from 'react';

export const createSystemLink = (id, name, keyData) => (
  <a
    rel="noreferrer"
    target="_blank"
    key={keyData}
    href={
      insights.chrome.isBeta()
        ? `/beta/insights/inventory/${id}`
        : `/insights/inventory/${id}`
    }
  >
    {name}
  </a>
);
