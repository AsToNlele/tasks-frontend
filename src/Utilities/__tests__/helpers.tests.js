import { camelCase, getProperty, orderArrayByProp, uniq } from '../helpers';
import items from '../hooks/useTableTools/__tests__/__fixtures__/items.fixtures';

describe('uniq', () => {
  it('should deduplicate items', () => {
    let result = uniq(['a', 'b', 'b', 'c']);
    expect(result.sort()).toEqual(['a', 'b', 'c']);
  });
});

describe('orderArrayByProp', () => {
  it('sorts an array by numbers asc', () => {
    expect(orderArrayByProp('system_count', items, 'asc')).toMatchSnapshot();
  });

  it('sorts an array by numbers desc', () => {
    expect(orderArrayByProp('system_count', items, 'desc')).toMatchSnapshot();
  });

  it('sorts an array by string asc', () => {
    expect(orderArrayByProp('title', items, 'asc')).toMatchSnapshot();
  });

  it('sorts an array by string desc', () => {
    expect(orderArrayByProp('title', items, 'desc')).toMatchSnapshot();
  });
});

/* stand-in test. may not need this function in tasks */
describe('getPropery', () => {
  const object = {
    level1: {
      level2: {
        level3: {
          value: 'value-level-3',
        },
        value: 'value-level-2',
      },
    },
  };

  it('should return values of keys', () => {
    expect(getProperty(object, 'level1.level2.level3.value')).toMatchSnapshot();
    expect(getProperty(object, 'level1.level2.value')).toMatchSnapshot();
    expect(getProperty({}, '', jest.fn())).toMatchSnapshot();
  });
});

describe('camelCase', () => {
  it('should set to camelcase', () => {
    expect(camelCase('system_count')).toBe('SystemCount');
  });
});
