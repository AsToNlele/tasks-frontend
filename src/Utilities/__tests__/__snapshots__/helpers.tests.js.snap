// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPropery should return values of keys 1`] = `"value-level-3"`;

exports[`getPropery should return values of keys 2`] = `"value-level-2"`;

exports[`getPropery should return values of keys 3`] = `[MockFunction]`;

exports[`orderArrayByProp sorts an array by numbers asc 1`] = `
Array [
  Object {
    "end": "2022-04-28T11:10:00",
    "id": 3,
    "initiated_by": "Michael",
    "run_date_time": "25 Apr 2022, 11:10 UTC",
    "start": "2022-04-25T10:10:00",
    "status": "complete",
    "system_count": 3,
    "title": "taskB",
  },
  Object {
    "end": "2022-04-20T11:10:00",
    "id": 2,
    "initiated_by": "Michael",
    "run_date_time": "20 Apr 2022, 11:10 UTC",
    "start": "2022-04-20T10:10:00",
    "status": "complete",
    "system_count": 5,
    "title": "taskB",
  },
  Object {
    "end": null,
    "id": 1,
    "initiated_by": "Michael",
    "run_date_time": "Running",
    "start": "2022-04-21T10:10:00",
    "status": "running",
    "system_count": 10,
    "title": "taskA",
  },
]
`;

exports[`orderArrayByProp sorts an array by numbers desc 1`] = `
Array [
  Object {
    "end": null,
    "id": 1,
    "initiated_by": "Michael",
    "run_date_time": "Running",
    "start": "2022-04-21T10:10:00",
    "status": "running",
    "system_count": 10,
    "title": "taskA",
  },
  Object {
    "end": "2022-04-20T11:10:00",
    "id": 2,
    "initiated_by": "Michael",
    "run_date_time": "20 Apr 2022, 11:10 UTC",
    "start": "2022-04-20T10:10:00",
    "status": "complete",
    "system_count": 5,
    "title": "taskB",
  },
  Object {
    "end": "2022-04-28T11:10:00",
    "id": 3,
    "initiated_by": "Michael",
    "run_date_time": "25 Apr 2022, 11:10 UTC",
    "start": "2022-04-25T10:10:00",
    "status": "complete",
    "system_count": 3,
    "title": "taskB",
  },
]
`;

exports[`orderArrayByProp sorts an array by string asc 1`] = `
Array [
  Object {
    "end": null,
    "id": 1,
    "initiated_by": "Michael",
    "run_date_time": "Running",
    "start": "2022-04-21T10:10:00",
    "status": "running",
    "system_count": 10,
    "title": "taskA",
  },
  Object {
    "end": "2022-04-20T11:10:00",
    "id": 2,
    "initiated_by": "Michael",
    "run_date_time": "20 Apr 2022, 11:10 UTC",
    "start": "2022-04-20T10:10:00",
    "status": "complete",
    "system_count": 5,
    "title": "taskB",
  },
  Object {
    "end": "2022-04-28T11:10:00",
    "id": 3,
    "initiated_by": "Michael",
    "run_date_time": "25 Apr 2022, 11:10 UTC",
    "start": "2022-04-25T10:10:00",
    "status": "complete",
    "system_count": 3,
    "title": "taskB",
  },
]
`;

exports[`orderArrayByProp sorts an array by string desc 1`] = `
Array [
  Object {
    "end": "2022-04-20T11:10:00",
    "id": 2,
    "initiated_by": "Michael",
    "run_date_time": "20 Apr 2022, 11:10 UTC",
    "start": "2022-04-20T10:10:00",
    "status": "complete",
    "system_count": 5,
    "title": "taskB",
  },
  Object {
    "end": "2022-04-28T11:10:00",
    "id": 3,
    "initiated_by": "Michael",
    "run_date_time": "25 Apr 2022, 11:10 UTC",
    "start": "2022-04-25T10:10:00",
    "status": "complete",
    "system_count": 3,
    "title": "taskB",
  },
  Object {
    "end": null,
    "id": 1,
    "initiated_by": "Michael",
    "run_date_time": "Running",
    "start": "2022-04-21T10:10:00",
    "status": "running",
    "system_count": 10,
    "title": "taskA",
  },
]
`;
